File: python/bib.py
import cv2 as cv
from PIL import Image
import os

class Matrice:
    def __init__(self, size):
        self.size = size
        self.matrice_content = None




    def getSize(self):
        return self.size

    def getPixel(self, x, y):

        return self.matrice_content[y][x]


    def getMatriceContent(self):
        return self.matrice_content


    def initContent(self, matrice):

        matrice_content = []


        y = 0
        while y < self.getSize():

            ligne = []

            x = 0
            while x < self.getSize():
                red = matrice[y][x][0]
                green = matrice[y][x][1]
                blue = matrice[y][x][2]
                color = (red, green, blue)

                pixel = Pixel(x, y, color)
                ligne.append(pixel)
                x +=1


            matrice_content.append(ligne)
            y+=1

        self.matrice_content = matrice_content


    def toStringColor(self):

        for line in self.matrice_content:
            print(self.getLineColor(line, 0))


    def getLineColor(self, line, pos):

        if pos < len(line):
            return str(line[pos].getColor()) + self.getLineColor(line, pos+1)

        else:
            return ""



    def toStringPixel(self):

        for line in self.matrice_content:
            print(self.getLinePixel(line, 0))


    def getLinePixel(self, line, pos):

        if pos < len(line):

            if line[pos].getColor() != (255, 255, 255):
                return " " + str(0) + " " + self.getLinePixel(line, pos+1)

            else:
                return " " + str(1) + " " + self.getLinePixel(line, pos+1)

        else:
            return ""






class Pixel:
    def __init__(self, x, y, color):
        self.x = x
        self.y = y
        self.color = color

    def getX(self):
        return self.x

    def getY(self):
        return self.y

    def getColor(self):
        return self.color

    def getColorRed(self):
        return self.color[0]

    def getColorGreen(self):
        return self.color[1]

    def getColorBlue(self):
        return self.color[2]


    def compare(self, pixel):

        return self.getColorRed() == pixel.getColorRed() and self.getColorGreen() == pixel.getColorGreen() and self.getColorBlue() == pixel.getColorBlue()





class Opencv:
    def __init__(self, nameExercice):
        self.nameExercice = nameExercice
        self.numberImage = 0
        self.listPatternInit = []
        self.nbMatriceResult = 0
        self.sizeImage = 0



    def setNumberImageResultat(self, nbMatriceResult):
        self.nbMatriceResult = nbMatriceResult

    def initSizeImage(self):
        img = Image.open("./pattern/"+self.nameExercice+"/option1.png")
        self.sizeImage = img.size[0]


    def getNumberImage(self):

        PATH = "./pattern/"+self.nameExercice

        cpt = 0
        for image in os.listdir(PATH):
            if image.endswith(".png"):
                cpt += 1

        self.numberImage = cpt - self.nbMatriceResult


    def extractImage(self):

        i = 1

        while i <= self.numberImage:
            img = cv.imread("./pattern/"+self.nameExercice+"/option"+str(i)+".png")
            i += 1

            self.listPatternInit.append(img)


    def initExercice(self, size_matrice):
        newListPattern=[]

        print(len(self.listPatternInit))

        for pattern in self.listPatternInit:
            newMatrice = Matrice(size_matrice)
            newMatrice.initContent(pattern)
            print(newMatrice.matrice_content)

            newListPattern.append(newMatrice)

        return newListPattern


    def matriceResult(self,size_matrice):

        img = cv.imread("./pattern/" + self.nameExercice + "/result.png")
        resultMatrice = Matrice(size_matrice)
        resultMatrice.initContent(img)

        return resultMatrice


class Exercice:

    def __init__(self, resultat, nameExercice):
        self.resultat = resultat
        self.nameExercice = nameExercice



    def assertRes(self,solution_user, resultat):
        if solution_user == resultat:
            return "SUCCESS"
        else:
            return "ERROR"





if __name__ == '__main__':
    opencv = Opencv("comparaisonMatrice")
    opencv.getNumberImage()
    print(opencv.numberImage)

File: cpp/bib.cpp
    #include <vector>
    #include <iostream>
    #include <string>
    #include <opencv2/highgui.hpp>
    #include <dirent.h>
    #include <sys/types.h>
    #include <filesystem>


    #include "Matrice.h"
    #include "Opencv.h"
    #include "Exercice.h"  
    #include "Line.h"
    #include "Pixel.h"
    using namespace std;





        Pixel::Pixel(int x, int y, int *color){
            this->x = x;
            this->y = y;
            this->color[0] = color[0];
            this->color[1] = color[1];
            this->color[2] = color[2];
        }

      

        int Pixel::getX(){
            return this->x;
        }
        
        int Pixel::getY(){
            return this->y;
        }


      
        int *Pixel::getColor(){
            return this->color;
        }
        
        int Pixel::getColorRed(){
            return this->color[0];
        }

    
        int Pixel::getColorGreen(){
            return this->color[1];
        }
        
        int Pixel::getColorBlue(){
            return this->color[2];
        }

        bool Pixel::compare(Pixel pixel2){

            return this->color[0] == pixel2.getColorRed() && this->color[1] == pixel2.getColorGreen() && this->color[2] == pixel2.getColorBlue();
        }



        Exercice::Exercice(int resultat, std::string nameExercice){
            this->resultat = resultat;
            this->nameExercice = nameExercice;
        }


        std::string Exercice::assertRes(int solution_user, int resultat){

            if (solution_user == resultat){
                return "SUCESS";
            }else{
                return "ERROR";
            }
        }

        std::string Exercice::assertResMult(std::vector<double> solution_user, std::vector<double> resultat){

            bool equal = true;

            if(solution_user.size() == resultat.size()){

                for(unsigned int i = 0; i < solution_user.size(); i++){
                    equal = equal && solution_user[i] == resultat[i];
                }

            }else{
                return "ERROR";
            }

            if (equal){
                return "SUCESS";
            }else{
                return "ERROR";
            }
        }


        std::string Exercice::assertResMultInt(std::vector<int> solution_user, std::vector<int> resultat){

            bool equal = true;

            if(solution_user.size() == resultat.size()){

                for(unsigned int i = 0; i < solution_user.size(); i++){
                    equal = equal && solution_user[i] == resultat[i];
                }

            }else{
                return "ERROR";
            }

            if (equal){
                return "SUCESS";
            }else{
                return "ERROR";
            }
        }




        Line::Line(){
        }

        int Line::getSize(){
            return this->lineContent.size();
        }


        void Line::insertPixel(Pixel p){
            this->lineContent.push_back(p);
        }


        Pixel Line::getPixelFromLine(int x){
            return this->lineContent[x];
        }

      
      void Line::printEachPixelValuesOfOneLine(){

            std::string stringLine = "";

            for(unsigned int i = 0; i < this->lineContent.size(); i++){

                Pixel pixel = this->lineContent[i];

                if (pixel.getColorBlue() != 255 && pixel.getColorGreen() != 255  && pixel.getColorRed() !=255 ){
                    stringLine = stringLine + "1 ";
                }else{
                    stringLine = stringLine + "0 ";
                }

            }

            std::cout << stringLine << std::endl;
        }


        void Line::printEachPixelsOfOneLine(){

            std::string stringLine = "";

            for(unsigned int i = 0; i < this->lineContent.size(); i++){
                Pixel pixel = this->lineContent[i];
                stringLine = stringLine + "("+std::to_string(pixel.getColorRed())+","+std::to_string(pixel.getColorGreen())+","+std::to_string(pixel.getColorBlue())+")";
            }

            std::cout << stringLine << std::endl;
        }







        Matrice::Matrice(int size){
            this->size = size;
        }


        int Matrice::getSize(){
            return this->size;
        }


        Pixel Matrice::getPixel(int x, int y){

                Line currentLine = this->matriceContent[y];

                Pixel pixel = currentLine.getPixelFromLine(x);
                return pixel;
            }

        std::vector<Line> Matrice::getMatriceContent(){

            return this->matriceContent;
        }



        void Matrice::toStringPixel(){

              for(unsigned int i = 0; i < this->matriceContent.size(); i++){
                Line line = this->matriceContent[i];
                line.printEachPixelValuesOfOneLine();
            }
        }

        void Matrice::toStringColor(){

              for(unsigned int i = 0; i < this->matriceContent.size(); i++){
                Line line = this->matriceContent[i];
                line.printEachPixelsOfOneLine();
            }
        }

   

        void Matrice::initContent(cv::Mat img){

            int y = 0;

          for(y=0; y < this->size; y++){

                Line line = Line();
                int x = 0;

                for(x=0; x<this->size; x++){


                    int red = img.at<cv::Vec3b>(y,x)[0];
                    int green = img.at<cv::Vec3b>(y,x)[1];
                    int blue = img.at<cv::Vec3b>(y,x)[2];
                    ;
                    int color[3] = {red, green, blue};

                    Pixel pixel = Pixel(x, y, color);
                    line.insertPixel(pixel);
                }

                this->matriceContent.push_back(line);

            }
        }







        Opencv::Opencv(std::string nameExercice){
            this->nameExercice = nameExercice;
        }


      int Opencv::getSizeImage(){
            return this->sizeImage;
            
        }

        void Opencv::setNumberImageResultat(int nbMatriceResult){
            this->nbMatriceResult = nbMatriceResult;
        }


        void Opencv::initSizeImage(){
            std::string image_path = cv::samples::findFile("./pattern/"+this->nameExercice+"/option1.png");
            cv::Mat img = cv::imread(image_path, cv::IMREAD_COLOR);
            this->sizeImage = img.size[0];
        }


        void Opencv::getNumberImage(){

            std::string path = "./pattern/"+this->nameExercice;
            int cptImage = 0;
            DIR *dir;
            struct dirent *ent;

            if ((dir = opendir (path.c_str())) != NULL) {
                cptImage = 0;
                while ((ent = readdir (dir)) != NULL) {

                    std::string name = ent->d_name;
                    std::string subStringName = name.substr(0, 6);    

                    if(subStringName=="option"){
                        cptImage++;
                    }

                }
                closedir (dir);
                this->numberImage = cptImage;
            } else {
                
            perror ("");
            }
            
        }

        void Opencv::extractImage(){
            int i = 1;
           

            while(i <= this->numberImage){
                std::string PATH = "./pattern/"+this->nameExercice+"/option"+std::to_string(i)+".png";
                std::string image_path = cv::samples::findFile(PATH);
                cv::Mat img = cv::imread(image_path);
                this->listPatternInit.push_back(img);
                i++;
              //  Matrice pattern = listPatternInit[i];
                //pattern.toStringPixel();
   

            }
        }


        Matrice Opencv::matriceResult(){

            std::string image_path = cv::samples::findFile("./pattern/"+this->nameExercice+"/result.png");
            cv::Mat img = cv::imread(image_path);
            Matrice matriceResult = Matrice(this->sizeImage);
            matriceResult.initContent(img);
            return matriceResult;

        }


        std::vector<Matrice> Opencv::initExercice(){
            std::vector<Matrice> listPattern;

            for(unsigned int i = 0; i < this->listPatternInit.size(); i++){
                cv::Mat mat = this->listPatternInit[i];
                Matrice matrice = Matrice(this->sizeImage);
                matrice.initContent(mat);
                listPattern.push_back(matrice);
            }

            return listPattern;

        }









